nexus/reconfigurator/planning/src/planner.rs

```diff
@@ -1401,27 +1402,65 @@ impl<'a> Planner<'a> {
         // For better or worse, switches and PSCs do not have the same idea of
         // being adopted into the control plane.  If they're present, they're
         // part of the system, and we will update them.
-        let included_sled_baseboards: BTreeSet<_> = self
+        let mut included_sled_baseboards: BTreeMap<_, _> = self
             .input
             .all_sleds(SledFilter::SpsUpdatedByReconfigurator)
-            .map(|(_sled_id, details)| &details.baseboard_id)
+            .map(|(sled_id, details)| (&details.baseboard_id, sled_id))
             .collect();
 
-        let included_baseboards =
-            self.inventory
-                .sps
-                .iter()
-                .filter_map(|(baseboard_id, sp_state)| {
-                    let do_include = match sp_state.sp_type {
-                        SpType::Sled => included_sled_baseboards
-                            .contains(baseboard_id.as_ref()),
-                        SpType::Power => true,
-                        SpType::Switch => true,
-                    };
-                    do_include.then_some(baseboard_id.clone())
+        // We only keep sled baseboards that do not contain zones that are
+        // unsafe to shut down
+        included_sled_baseboards.retain(|baseboard_id, sled_id| {
+            let zones = self.blueprint.current_sled_zones(
+                *sled_id,
+                BlueprintZoneDisposition::is_in_service,
+            );
+
+            let unsafe_zones: Vec<_> = zones
+                .into_iter()
+                .filter(|zone| {
```

Comment:
Nit - we collect a vec of blueprint zone configs here, then collect another vec of zone kind strings below. I think we could just collect one if we made this one collect the strings, either by adding a `.map(|zone| zone.kind.report_str())` after filtering or by changing `filter` to `filter_map`?